src/lib/zork/index.ts

```diff
@@ -272,7 +292,47 @@ const onDataChannel = (ctx: Context, channel: any) => {
   // it forwards data for the SOCKS session.
   channel.onopen = () => ctx.log(`[channel ${channelId}] datachannel opened`);
   initSocksSessionAndFwdSocket(ctx, channel);
-  channel.onmessage = (event: any) => ctx.socksSession.handleDataFromSocksClient(event.data);
+  channel.onmessage = (event: any) => {
+    let bytesForSession = event.data;
+
+    // only legacy clients send us string messages.
+    if (typeof event.data === 'string') {
+      // this is done by peerconnection.ts - a one word string ('heartbeat')
+      if (event.data === HEARTBEAT_MSG) {
+        channel.send(HEARTBEAT_MSG);
+        return;
+      }
+
+      const request = JSON.parse(event.data);
+      if (request.control) {
+        // pool.ts
+        const command = request.control;
+        switch (command) {
+          case 'OPEN':
+            ctx.log('pool OPEN - resetting session!');
+            initSocksSessionAndFwdSocket(ctx, channel);
+            break;
+          case 'CLOSE':
+            ctx.log('ignoring pool CLOSE');
+            break;
+          default:
+            ctx.log(`unknown pool command ${command}`);
+        }
+        return;
+      }
+
+      try {
+        // pool.ts adds an additional layer of wrapping.
+        // the actual message is in a field called data, again encoded as JSON.
+        bytesForSession = socks_headers.composeRequestBuffer(<socks_headers.Request>JSON.parse(request.data));
+      } catch (e) {
+        ctx.log(`cannot parse legacy endpoint: ${event.data} (${e.message})`);
+        // TODO: fail!
+      }
+    }
+
+    ctx.socksSession.handleDataFromSocksClient(bytesForSession);
+  };
```

Comment:
#2859 removes ctx.socksSession in favor of a local socks session variable. If that's merged first, I'm happy to resolve the resulting conflicts with this change.